x/config/configurator.go

```
@@ -0,0 +1,230 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/internal/decode"
+	errs "go.uber.org/yarpc/internal/errors"
+
+	"gopkg.in/yaml.v2"
+)
+
+// Configurator helps build Dispatchers using runtime configuration.
+//
+// An empty Configurator does not know about any transports. Inform it about
+// the different transports and their configuration parameters using the
+// RegisterTransport function.
+type Configurator struct {
+	knownTransports map[string]*compiledTransportSpec
+}
+
+// New sets up a new empty Configurator. The returned Configurator does not
+// know about any transports. Individual TransportSpecs must be registered
+// against it using the RegisterTransport function.
+func New() *Configurator {
+	return &Configurator{knownTransports: make(map[string]*compiledTransportSpec)}
+}
+
+// RegisterTransport registers a TransportSpec with the given Configurator. An
+// error is returned if the TransportSpec was invalid.
+//
+// If a transport with the same name was already registered, it will be
+// overwritten.
+//
+// Use MustRegisterTransport if you want to panic in case of registration
+// failure.
+func (c *Configurator) RegisterTransport(t TransportSpec) error {
+	if t.Name == "" {
+		return errors.New("name is required")
+	}
+
+	spec, err := compileTransportSpec(&t)
+	if err != nil {
+		return fmt.Errorf("invalid TransportSpec for %q: %v", t.Name, err)
+	}
+
+	// TODO: Panic if a transport with the given name is already registered?
```

can we return an error for now?